Handle cmark parsing options

tvon opened the pull request:
This adds support for the [CMARK_OPT_*][1] flags.

The flags are defined as:

* `default`: No options, provided to align with the cmark source.
  Passing `default: true` does nothing, and any other flags will
  override it due to how the option value is calculated (bitwise).
* `sourcepos`: Include a `data-sourcepos` attribute on all block elements
* `hardbreaks`: Render `softbreak` elements as hard line breaks
* `safe`: Suppress raw HTML and unsafe links (`javascript:`, `vbscript:`,
  `file:`, and `data:`, except for `image/png`, `image/gif`,
  `image/jpeg`, or `image/webp` mime types).  Raw HTML is replaced
  by a placeholder HTML comment. Unsafe links are replaced by
  empty strings.
* `normalize`: Normalize tree by consolidating adjacent text nodes.
* `validate_utf8`: Validate UTF-8 in the input before parsing, replacing
  illegal sequences with the replacement character U+FFFD.
* `smart`: Convert straight quotes to curly, --- to em dashes, -- to en dashes.

Options are enabled by passing a list to `to_html/2` or `to_html/3`,
e.g.:

```elixir
Cmark.to_html(some_markdown_string, [smart: true, sourcepos: true])
```

TODO:
- [x] Review C code, because I have no idea.
- [x] Update exdoc/README as appropriate.
- [x] ~~Is passing a list of booleans with true/false appropriate or
  should it just be a list of symbols? E.g. `[smart: true, sourcepos:
  true]` vs just `[:smart, :sourcepos]`.~~ Rewrite params as keyword, list, e.g. `[:smart, :sourcepos]`
- [x] ~~Should `@flags` be a module attribute or just a map in the method?~~ Move `@flags` into method
- [x] Review method of parsing/calculating bitwise options value.
- [x] Review function signatures/definitions, seems kind of messy.
- [x] Write some more specific tests.  The tests generated from cmark don't
  exercise any of the option flags beyond `smart`.

[1]: https://github.com/jgm/cmark/blob/0.22.0/src/cmark.h#L450-L481

tvon commented:
@asaaki This needs some more work but I would appreciate any feedback you have.  Also note that I'm new to elixir so  this is probably not very idiomatic at this point.

asaaki commented:
First of all: Thank you for your contribution!

Second: Don't worry. I'm not an expert at all, just a curious coder.

Third: Thank you again, especially for this feature. When I initially built this package I focused only on the default, since this is what most people probably will use most of the time. I had the options in my mind, but never felt the urge to actually implement it.

I'll have a look at this PR within the next days. The general layout seems to be quite good. I guess we just need to polish a thing here and there.

Regarding the options dict/list: since the default is to have nothing activated and we just have to deal with opt-in switches, I'd prefer a `[:list, :of, :atoms]`.

Also:
Happy holidays! :christmas_tree:

tvon commented:
@asaaki Okay, I think it's in pretty good shape now.

* Changed options to list of atoms and moved to last param of each function.
* Added simple doctest examples where appropriate.
* The C changes are pretty minor so I think they're probably okay.
* The Makefile and test/cmark_test.exs changes aren't the best, but they get the job done.  I'm open to any suggestions here.

tvon commented on lib/cmark.ex:
```diff
@@ -56,31 +122,47 @@ defmodule Cmark do
+      safe: 32
+    }
+
+    Enum.reduce(options, 0, fn(x, acc) -> flags[x] ||| acc end)
```
I'm not sure how idiomatic this function is.

asaaki replied:
With the change suggestion from above (and a tiny renaming):

```elixir
Enum.reduce(options, 0, fn(flag, acc) -> @flags[flag] + acc end)
```

I did some benchmarking and noticed that `Bitwise` does not significantly improve the performance (both approaches are on par), so I vote for the slightly simpler `+` here, as it makes the code more understandable.

tvon replied:
Yeah, that is a great change.

I also am not sure why I used `Bitwise` at all.

asaaki commented on lib/cmark.ex:
```diff
@@ -24,11 +25,50 @@ defmodule Cmark do
-  def to_html(data) when is_list(data),      do: parse_doc_list(data)
-  def to_html(data) when is_bitstring(data), do: parse_doc(data)
+  def to_html(data) when is_list(data) do
+    parse_doc_list(data, [:default])
```
We can remove the `default` handling, as it would result in `0` anyway.

asaaki commented on lib/cmark.ex:
```diff
@@ -24,11 +25,50 @@ defmodule Cmark do
+  end
+
+  def to_html(data) when is_bitstring(data) do
+    parse_doc(data, [:default])
```
See comment above.

asaaki commented on lib/cmark.ex:
```diff
@@ -24,11 +25,50 @@ defmodule Cmark do
+  Compiles one or more (list) Markdown documents to HTML using provided options
+  and returns result.
+
+  Options are paseed as a list of atoms.  Available options are:
```
Typo: paseed -> passed

asaaki commented on lib/cmark.ex:
```diff
@@ -43,11 +83,37 @@ defmodule Cmark do
-  def to_html(data, callback) when is_list(data),      do: parse_doc_list(data, callback)
-  def to_html(data, callback) when is_bitstring(data), do: parse_doc(data, callback)
+  def to_html(data, callback) when is_list(data) and is_function(callback) do
+    parse_doc_list(data, callback, [:default])
```
See comments above.

asaaki commented on lib/cmark.ex:
```diff
@@ -43,11 +83,37 @@ defmodule Cmark do
+  end
+
+  def to_html(data, callback) when is_bitstring(data) and is_function(callback) do
+    parse_doc(data, callback, [:default])
```
See comments above.

asaaki commented on lib/cmark.ex:
```diff
@@ -56,31 +122,47 @@ defmodule Cmark do
   """
-  def to_html_each(data, callback) when is_list(data) do
-    parse_doc_list_each(data, callback)
+  def to_html_each(data, callback, options \\ [:default]) when is_list(data) do
```
Change `options \\ [:default]` to `options \\ []` (see comments above).

asaaki commented on lib/cmark.ex:
```diff
@@ -56,31 +122,47 @@ defmodule Cmark do
+  defp parse_options(options) do
+    use Bitwise
+
+    flags = %{
```
I would convert this into a module attribute and a simple dict (maps are cool, but not necessary here):

```elixir
@flags [
  sourcepos: 1,
  hardbreaks: 2,
  normalize: 4,
  smart: 8,
  validate_utf8: 16,
  safe: 32
]
```

`default` removed, as it would not change the result. Also see comments above (about not passing this key).

asaaki replied:
I realized that you had this thing in your TODO list.
What was the reason for moving it into the method?
For static data like this map/dict I would always vote for a module attribute.

tvon replied:
No good reason, just some elixir ignorance.  I wasn't sure if a module attribute would be documented or not and I figured it was a private map so it didn't need to be exposed.

tvon replied:
Is a dict just `foo = [ping: 1]`?

tvon replied:
Ah, yes, I see you did the conversion in your comment.

asaaki commented:
Added feedback to some places.
In general this looks pretty good now.
After the fixes/changes I would say it's ready for merging.

asaaki commented:
Also marked the `C code` item from the TODO, since there was no big change and it looks fine to me (and since the test passed, the compiler thinks the same :smile:).

tvon commented:
@asaaki What do you think of this?

I removed the default `options` handling in `parse_doc*` because it seemed to make things unnecessarily messy with regards to function signatures clashing.  As it stands the public functions will pass `[]` instead of expecting the `parse_doc*` functions to handle whatever is thrown at them.

asaaki commented:
I think it looks totally fine to me.
As far as I can tell the interface looks clean and neat to me.

If you have no objections I would merge it soon after I got some rest. :smile: 
(And prepare a new release, of course, so you can use the new version pretty soon.)

Happy New Year from Germany! :fireworks:

tvon commented:
Awesome, thanks for your time.  Happy New Year from Ohio, USA!

tvon commented:
squashed
